### PR TITLE
[WIP] edit error handling for object with status and response

### PR DESCRIPTION
**Pull request by mcbouslog**
Not ready for merge, creating for discussion.

This is the 2nd PR to be merged in relation to improving API error handling.
1st to be merged would be the [related PR in `json-api-client`](https://github.com/zooniverse/json-api-client/pull/37), which contains more context and questions.

### CODE REVIEW

File: lib/api-client.js [hunk 1]
```diff
@@ -6,22 +6,21 @@ var apiClient = new JSONAPIClient(config.host + '/api', {
   'Accept': 'application/vnd.api+json; version=1',
 }, {
   beforeEveryRequest: function() {
+    console.log('using local panoptes-client');
     var auth = require('./auth');
     return auth.checkBearerToken();
   },
 
-  handleError: function(response) {
+  handleError: function(error) {
```
Comment (mcbouslog) [inline lib/api-client.js]: if `error` is `undefined` I need to create it as a new Error object so an `error.message` can be added.

File: lib/api-client.js [hunk 2]
```diff
@@ -6,22 +6,21 @@ var apiClient = new JSONAPIClient(config.host + '/api', {
   'Accept': 'application/vnd.api+json; version=1',
 }, {
   beforeEveryRequest: function() {
+    console.log('using local panoptes-client');
```
Comment (mcbouslog) [inline lib/api-client.js]: will remove, keeping momentarily while try to stage branch of PFE with dependency as GitHub branch that itself has a dependency as GitHub branch.

File: lib/api-client.js [hunk 3]
```diff
@@ -33,19 +32,25 @@ var apiClient = new JSONAPIClient(config.host + '/api', {
       } else {
         errorMessage = 'Unknown error (bad response body)';
       }
-    } else if (response.text.indexOf('<!DOCTYPE') !== -1) {
+    } else if (error.response.text.indexOf('<!DOCTYPE') !== -1) {
       // Manually set a reasonable error when we get HTML back (currently 500s will do this).
```
Comment (eatyourgreens) [inline lib/api-client.js]: If we have the status code in the response, why are we sniffing the response body to test for a 500 error?

File: lib/api-client.js [hunk 4]
```diff
@@ -33,19 +32,25 @@ var apiClient = new JSONAPIClient(config.host + '/api', {
...
 
-    throw new Error(errorMessage);
+    if (typeof error === 'undefined') {
+      error = new Error(errorMessage);
+    } else if (typeof error === 'object') {
+      error.message = errorMessage;
+    }
+
+    throw error;
   }
```
Comment (eatyourgreens) [inline lib/api-client.js]: Can you call throw with something that isn't an instance of Error?

File: lib/api-client.js [hunk 5]
```diff
@@ -6,22 +6,21 @@ var apiClient = new JSONAPIClient(config.host + '/api', {
...
     var errorMessage;
-    if (response instanceof Error) {
-      throw response;
-    } else if (typeof response.body === 'object') {
-      if (response.body.error) {
-        errorMessage = response.body.error;
-        if (response.body.error_description) {
-          errorMessage += ' ' + response.error_description;
+    if (typeof error.response.body === 'object') {
+      if (error.response.body.error) {
```
Comment (eatyourgreens) [inline lib/api-client.js]: `error.response.body.error` is pretty convoluted, and confusing, here.

Comment (eatyourgreens) [inline lib/api-client.js]: Errors should maybe follow the [JSON API spec for error objects](http://jsonapi.org/format/1.1/#error-objects).

**Pull request closed without merging**